rclcpp/src/rclcpp/typesupport_helpers.cpp

```diff
@@ -0,0 +1,145 @@
+// Copyright 2018, Bosch Software Innovations GmbH.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "rclcpp/typesupport_helpers.hpp"
+
+#include <memory>
+#include <sstream>
+#include <stdexcept>
+#include <string>
+#include <tuple>
+#include <utility>
+
+#include "ament_index_cpp/get_package_prefix.hpp"
+#include "ament_index_cpp/get_resources.hpp"
+#include "rcpputils/shared_library.hpp"
+#include "rosidl_runtime_cpp/message_type_support_decl.hpp"
+
+namespace rclcpp
+{
+
+std::string get_typesupport_library_path(
+  const std::string & package_name, const std::string & typesupport_identifier)
+{
+  const char * filename_prefix;
+  const char * filename_extension;
+  const char * dynamic_library_folder;
+#ifdef _WIN32
+  filename_prefix = "";
+  filename_extension = ".dll";
+  dynamic_library_folder = "/bin/";
+#elif __APPLE__
+  filename_prefix = "lib";
+  filename_extension = ".dylib";
+  dynamic_library_folder = "/lib/";
+#else
+  filename_prefix = "lib";
+  filename_extension = ".so";
+  dynamic_library_folder = "/lib/";
+#endif
```

Comment:
It seems like we must have logic like this already elsewhere?

Reply:
I wonder if this logic could be used some how?

https://github.com/ros2/rcpputils/blob/71ceb6b0597eee71f91dfeffe77fca7cc1a9cdb4/src/find_library.cpp#L37-L52

Reply:
If you prefer, I can place it in a namespace in https://github.com/ros2/rcpputils/blob/master/include/rcpputils/find_library.hpp – on the other hand, duplication is not terrible here since these strings won't ever change and get out of sync.

Reply:
"won't ever get out of sync" is always said before things do get out of sync - code has to last a long time sometimes. The `find_library` function works fine, though it is a wider search scope than this helper function, I think a version with a specific search directory would be a nice addition to `rcpputils`.

Reply:
> "won't ever get out of sync" is always said before things do get out of sync - code has to last a long time sometimes. The find_library function works fine, though it is a wider search scope than this helper function, I think a version with a specific search directory would be a nice addition to rcpputils.

I agree. I won't block this pull request on that, but I will say the chances of a new OS with new values being added in one place but not the other seems very likely. And even getting out of sync due to a bug fix or something is possible. DRY applies here I think.

Reply:
I would highly recommend factoring this into a common function. Even if just gave the information containing in these `#ifdef` blocks.

Reply:
I thought I had mention this somewhere on this PR already, but I actually don't see this `typesupport_helpers` file suited within rclcpp. I think this should be a mix of `rosidl_runtime_cpp` and `rcpputils`. The first package has already a function called `get_message_type_support_handle()` (see https://github.com/ros2/rosidl/blob/master/rosidl_runtime_cpp/include/rosidl_runtime_cpp/message_type_support_decl.hpp#L30) and I could see that we introduce a function overlay which loads the type support given a package and message name. 

I'd be ok though with the changes as proposed with @wjwwood

Reply:
So should I move the generic parts into rcpputils in this PR?

Reply:
> So should I move the generic parts into `rcpputils` in this PR?

I don't think there's anything to move there, but there's logic in `rcpputils` that could replace logic in this pull request, if just adjusted.

We could do that in a follow up pr though. If you open an issue after this merged that would be good enough with me, maybe a TODO in the comments here.

Reply:
I used `rcpputils::kSolibPrefix` and `rcpputils::kSolibExtension`, but that's the extent to which code sharing with rcpputils makes sense in my opinion.`find_library_path` does something different from `get_typesupport_library_path` so generalization of the two into one function would be very ugly.

Reply:
How long will it take until CI picks up on the change I made in rcpputils after merging that change?

Reply:
Once the change is merged, someone needs to do an `rcpputils` release into Rolling.  That will rebuild everything above it, at which point we can re-run CI here.

Reply:
Usually I don't wait for that, instead if ci.ros2.org's CI is green I'd merge both, ignoring the automatically triggered ci from build.ros2.org.